File: slackblocks/rich_text.py
from abc import ABC, abstractmethod
from enum import Enum
from json import dumps
from typing import Any, Dict, List, Optional, TypeVar, Union

from .errors import InvalidUsageError
from .utils import coerce_to_list, validate_int


class RichTextObjectType(Enum):
    SECTION = "rich_text_section"
    LIST = "rich_text_list"
    PREFORMATTED = "rich_text_preformatted"
    QUOTE = "rich_text_quote"


class ListType(Enum):
    BULLET = "bullet"
    ORDERED = "ordered"

    def all() -> List[str]:
        return [list_type.value for list_type in ListType]


class RichTextObject(ABC):
    def __init__(self, type_: RichTextObjectType) -> None:
        self.type_ = type_

    @abstractmethod
    def _resolve(self) -> Dict[str, Any]:
        return {"type": self.type_.value}

    def __repr__(self) -> str:
        return dumps(self._resolve(), indent=4)


class RichTextSection(RichTextObject):
    def __init__(self, elements: Union[RichTextObject, List[RichTextObject]]) -> None:
        super().__init__(type_=RichTextObjectType.SECTION)
        self.elements = coerce_to_list(elements, class_=RichTextObject)

    def _resolve(self) -> Dict[str, Any]:
        section = super()._resolve()
        section["elements"] = [element._resolve() for element in self.elements]
        return section


class RichTextList(RichTextObject):
    def __init__(
        self,
        style: Union[str, ListType],
        elements: Union[RichTextSection, List[RichTextSection]],
        indent: Optional[int] = None,
        offset: Optional[int] = 0,
        border: Optional[int] = 0,
    ) -> None:
        super().__init__(type_=RichTextObjectType.LIST)
        if isinstance(style, str):
            if style in ListType.all():
                self.style = style
            else:
                raise InvalidUsageError(f"`style` must be one of [{ListType.all()}]")
        elif isinstance(style, ListType):
            self.style = style.value
        self.elements = coerce_to_list(elements, RichTextSection, min_size=1)
        self.indent = validate_int(indent, allow_none=True)
        self.offset = validate_int(offset, allow_none=True)
        self.border = validate_int(border, allow_none=True)

    def _resolve(self) -> Dict[str, Any]:
        rich_text_list = super()._resolve()
        rich_text_list["elements"] = [element._resolve() for element in self.elements]
        rich_text_list["style"] = self.style
        if self.indent is not None:
            rich_text_list["indent"] = self.indent
        if self.offset is not None:
            rich_text_list["offset"] = self.offset
        if self.border is not None:
            rich_text_list["border"] = self.border
        return rich_text_list


class RichTextSubElement(Enum):
    TEXT = "text"
    LINK = "link"


class RichText(RichTextObject):
    def __init__(
        self,
        text: str,
        bold: Optional[bool] = None,
        italic: Optional[bool] = None,
        strike: Optional[bool] = None,
    ) -> None:
        super().__init__(type_=RichTextSubElement.TEXT)
        self.text = text
        self.bold = bold
        self.italic = italic
        self.strike = strike

    def _resolve(self) -> Dict[str, Any]:
        rich_text = super()._resolve()
        rich_text["text"] = self.text
        style = {}
        if self.bold is not None:
            style["bold"] = self.bold
        if self.italic is not None:
            style["italic"] = self.italic
        if self.strike is not None:
            style["strike"] = self.strike
        if style:
            rich_text["style"] = style
        return rich_text


class RichTextLink(RichTextObject):
    def __init__(
        self,
        url: str,
        text: Optional[str] = None,
        unsafe: Optional[bool] = None,
        bold: Optional[bool] = None,
        italic: Optional[bool] = None,
    ) -> None:
        super().__init__(type_=RichTextSubElement.LINK)
        self.url = url
        self.text = text
        self.unsafe = unsafe
        self.bold = bold
        self.italic = italic

    def _resolve(self) -> Dict[str, Any]:
        link = super()._resolve()
        link["url"] = self.url
        if self.text is not None:
            link["text"] = self.text
        if self.unsafe is not None:
            link["unsafe"] = self.unsafe
        style = {}
        if self.bold is not None:
            style["bold"] = self.bold
        if self.italic is not None:
            style["italic"] = self.italic
        if style:
            link["style"] = style
        return link


class RichTextPreformattedCodeBlock(RichTextObject):
    def __init__(
        self,
        elements: Union[RichText, RichTextLink, List[Union[RichText, RichTextLink]]],
        border: Optional[int] = None,
    ) -> None:
        super().__init__(type_=RichTextObjectType.PREFORMATTED)
        self.elements = coerce_to_list(elements, (RichText, RichTextLink))
        self.border = border

    def _resolve(self) -> Dict[str, Any]:
        preformatted = super()._resolve()
        preformatted["elements"] = [element._resolve() for element in self.elements]
        if self.border is not None:
            preformatted["border"] = self.border
        return preformatted


class RichTextQuote(RichTextObject):
    def __init__(
        self,
        elements: Union[RichTextObject, List[RichTextObject]],
        border: Optional[int] = None,
    ) -> None:
        super().__init__(RichTextObjectType.QUOTE)
        self.elements = coerce_to_list(elements, RichTextObject)
        self.border = border

    def _resolve(self) -> Dict[str, Any]:
        quote = super()._resolve()
        quote["elements"] = [element._resolve() for element in self.elements]
        if self.border is not None:
            quote["border"] = self.border
        return quote


RichTextElement = TypeVar(
    "RichTextElement",
    RichTextList,
    RichTextPreformattedCodeBlock,
    RichTextQuote,
    RichTextSection,
)

File: typescript/src/legacy/rich-text.ts
/**
 * Inline and layout factories for Slack rich-text blocks.
 *
 * Build inline elements first, combine them in a section, list, quote, or code
 * block, then pass those layout objects to `richTextBlock`.
 *
 * @module rich-text
 */
import { create } from "../internal.js";
import type { FactorySettings, JsonObject } from "../types.js";

/** Inline formatting supported by rich-text text, links, users, and channels. */
export interface RichTextStyle {
  /** Render the inline content in bold. */
  bold?: boolean;
  /** Render the inline content in italics. */
  italic?: boolean;
  /** Render the inline content with a strikethrough. */
  strike?: boolean;
  /** Render the inline content as code. */
  code?: boolean;
}

/**
 * Creates a styled rich-text text run.
 *
 * @param text - Text content.
 * @param style - Optional inline formatting.
 * @param settings - Per-call validation settings.
 * @returns A validated Slack rich-text `text` element.
 */
export function richText(
  text: string,
  style?: RichTextStyle,
  settings: FactorySettings = {},
) {
  return create("text", { text, style }, settings);
}

/**
 * Creates a rich-text channel mention.
 *
 * @param channelId - Slack channel identifier.
 * @param style - Optional inline formatting.
 * @param settings - Per-call validation settings.
 * @returns A validated Slack rich-text `channel` element.
 */
export function richTextChannel(
  channelId: string,
  style?: RichTextStyle,
  settings: FactorySettings = {},
) {
  return create("channel", { channelId, style }, settings);
}

/**
 * Creates a rich-text emoji.
 *
 * @param name - Emoji name without surrounding colons.
 * @param settings - Per-call validation settings.
 * @returns A validated Slack rich-text `emoji` element.
 */
export function richTextEmoji(
  name: string,
  settings: FactorySettings = {},
) {
  return create("emoji", { name }, settings);
}

/**
 * Creates a rich-text link.
 *
 * @param input - Destination URL, optional label, safety flag, and formatting.
 * @param settings - Per-call validation settings.
 * @returns A validated Slack rich-text `link` element.
 */
export function richTextLink(
  input: {
    /** Destination URL. */
    url: string;
    /** Optional visible label. Slack displays the URL when omitted. */
    text?: string;
    /** Mark a URL as unsafe when mirroring a Slack-provided payload. */
    unsafe?: boolean;
    /** Optional inline formatting. */
    style?: RichTextStyle;
  },
  settings: FactorySettings = {},
) {
  return create("link", input, settings);
}

/**
 * Creates a rich-text user mention.
 *
 * @param userId - Slack user identifier.
 * @param style - Optional inline formatting.
 * @param settings - Per-call validation settings.
 * @returns A validated Slack rich-text `user` element.
 */
export function richTextUser(
  userId: string,
  style?: RichTextStyle,
  settings: FactorySettings = {},
) {
  return create("user", { userId, style }, settings);
}

/**
 * Creates a rich-text user-group mention.
 *
 * @param usergroupId - Slack user-group identifier.
 * @param style - Optional inline formatting.
 * @param settings - Per-call validation settings.
 * @returns A validated Slack rich-text `usergroup` element.
 */
export function richTextUserGroup(
  usergroupId: string,
  style?: RichTextStyle,
  settings: FactorySettings = {},
) {
  return create("usergroup", { usergroupId, style }, settings);
}

/**
 * Creates a paragraph-like rich-text section.
 *
 * @param elements - Inline rich-text elements in display order.
 * @param settings - Per-call validation settings.
 * @returns A validated Slack `rich_text_section` object.
 */
export function richTextSection(
  elements: JsonObject[],
  settings: FactorySettings = {},
) {
  return create("rich_text_section", { elements }, settings);
}

/**
 * Creates an ordered or bulleted rich-text list.
 *
 * @param input - List items, marker style, and optional list layout.
 * @param settings - Per-call validation settings.
 * @returns A validated Slack `rich_text_list` object.
 * @throws MissingRequiredError when `style` is not provided.
 */
export function richTextList(
  input: {
    /** Rich-text section objects used as list items. */
    elements: JsonObject[];
    /** List marker style. */
    style: "bullet" | "ordered";
    /** Nesting depth. */
    indent?: number;
    /** Starting number for an ordered list. */
    offset?: number;
    /** Optional border thickness. */
    border?: number;
  },
  settings: FactorySettings = {},
) {
  return create("rich_text_list", input, settings);
}

/**
 * Creates a preformatted rich-text code block.
 *
 * @param elements - Inline rich-text elements displayed as code.
 * @param options - Optional border thickness.
 * @param settings - Per-call validation settings.
 * @returns A validated Slack `rich_text_preformatted` object.
 */
export function richTextCodeBlock(
  elements: JsonObject[],
  options: {
    /** Optional border thickness. */
    border?: number;
  } = {},
  settings: FactorySettings = {},
) {
  return create("rich_text_preformatted", { elements, ...options }, settings);
}

/**
 * Creates a rich-text block quote.
 *
 * @param elements - Inline rich-text elements displayed in the quote.
 * @param options - Optional border thickness.
 * @param settings - Per-call validation settings.
 * @returns A validated Slack `rich_text_quote` object.
 */
export function richTextQuote(
  elements: JsonObject[],
  options: {
    /** Optional border thickness. */
    border?: number;
  } = {},
  settings: FactorySettings = {},
) {
  return create("rich_text_quote", { elements, ...options }, settings);
}
